_drafts/2024-05-18-get-started-with-aws-tf-01.md

```diff
@@ -0,0 +1,254 @@
+}
+```
+
+Next, apply the security group to the ec2 instance, adding the following property to the ec2 instance:
+
+```terraform 
+# /infra/unicorn-api.tf
+
+resource "aws_instance" "web_service_vm" {
+
+    # ...
+
+  vpc_security_group_ids = [aws_security_group.unicorn_api_security_group.id]
+}
+```
+
+With these new resources defined, we can re-apply our changes with `terraform apply`. This will rebuild the EC2 instance, so we'll need to get the latest ip address so we can ping it.  
+Once the resources have been updated and the latest IP address copied, we can now ping our EC2 instance with:  
+
+`ping <the-ip-address>`
+
+## Step 4 - SSH into the EC2 Instance
+
+In this step we're going to create an ssh key pair using terraform. The public key will be added to the ec2 instance as one of its known ssh public keys.  
+We'll save the private key as a local file so we can reference it when we try connect to the ec2 instance.
+
+First, add the following providers to our `main.tf`:
+
+```terraform
+# infra/main.tf
+
+required_providers {
+
+    # ...
+
+    tls = {
+          source  = "hashicorp/tls"
+          version = "~> 4.0.4"
+        }
+
+    local = {
+      source  = "hashicorp/local"
+      version = "2.4.1"
+    }
+}
+```
+
+Since we've added new providers, we need to run `terraform init` to download their modules.  
+
+Create a new file named `ssh_key.tf` and add the following:
+
+```terraform
+# infra/ssh_key.tf
+
+# create a tls key pair using rsa algorithm
+resource "tls_private_key" "ssh_key" {
+    algorithm = "RSA"
+    rsa_bits = 4096
+}
+
+# save the private key to a local file named "key.pem"
+resource "local_file" "rsa_key" {
+    content = tls_private_key.ssh_key.private_key_openssh
+    filename = "${path.module}/key.pem"
+}
+
+# define an aws key pair, providing the public key from the tls key pair defined above
+resource "aws_key_pair" "key_pair_for_ec2_instance" {
+    key_name = "ssh-key"
+    public_key = tls_private_key.ssh_key.public_key_openssh
+}
+```
+
+Next, add the following property to your ec2 instance:
+
+```terraform
+# infra/unicorn-api.tf
+
+resource "aws_instance" "web_service_vm" {
+
+    # ...
+
+  key_name = aws_key_pair.key_pair_for_ec2_instance.key_name
+}
+```
+With those resources defined and linked to our EC2 instance, we can once again reapply our terraform config to deploy the changes to aws.
+
+Now that those changes have been applied we should now have a new local file, `key.pem`. Before we try ssh into our instance, we need to set the access permissions for this file. Something like `chmod 400 key.pem` should suffice. This will give the current account read permissions for the file, and no permissions for other users.  
+For more information on how chmod works, [see here](https://ss64.com/bash/chmod.html).  
+
+Next we can find the IPv4 DNS name of the instance from the aws management console, this will be something `ec2-11-11-11-11.your-aws-region.compute.amaonaws.com`.  
```

Comment:
It might be worth adding an [output value](https://developer.hashicorp.com/terraform/language/values/outputs) to your terraform code containing the EC2 instance's IP address to make it easier for readers to fetch after each change. eg

```terraform
output "ec2_public_ip" {
  description = "Public IP address of the EC2 instance"
  value       = aws_instance.web_service_vm.public_ip
}
```

Outputs are shown at the end of every apply, or alternatively you can fetch them with the `terraform output` command.

```shell
terraform output                # shows all outputs
terraform output ec2_public_ip  # will show just the output named "ec2_public_ip"
```